sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightStructureViewFactory.kt

```diff
@@ -0,0 +1,94 @@
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.intellij.lang
+
+import com.alecstrong.sqlite.psi.core.psi.SqliteTypes
+import com.intellij.ide.structureView.StructureViewModel
+import com.intellij.ide.structureView.StructureViewModelBase
+import com.intellij.ide.structureView.StructureViewTreeElement
+import com.intellij.ide.structureView.TreeBasedStructureViewBuilder
+import com.intellij.ide.util.treeView.smartTree.TreeElement
+import com.intellij.lang.PsiStructureViewFactory
+import com.intellij.navigation.ItemPresentation
+import com.intellij.navigation.NavigationItem
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.util.Iconable
+import com.intellij.pom.Navigatable
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.squareup.sqldelight.core.SqlDelightTypes
+import com.squareup.sqldelight.core.lang.SqlDelightFile
+import com.squareup.sqldelight.core.lang.util.childrenOfType
+import javax.swing.Icon
+
+class SqlDelightStructureViewFactory : PsiStructureViewFactory {
+  override fun getStructureViewBuilder(psiFile: PsiFile) = object : TreeBasedStructureViewBuilder() {
+    override fun createStructureViewModel(editor: Editor?) = SqlDelightStructureViewModel(psiFile)
+    override fun isRootNodeShown() = false
+  }
+}
+
+internal class SqlDelightStructureViewModel(
+  psiFile: PsiFile
+) : StructureViewModelBase(psiFile, SqlDelightStructureViewElement(psiFile)),
+    StructureViewModel.ElementInfoProvider {
+  override fun isAlwaysLeaf(element: StructureViewTreeElement) = false
+  override fun isAlwaysShowsPlus(element: StructureViewTreeElement) = false
+}
+
+internal class SqlDelightStructureViewElement(
+  private val element: PsiElement
+) : StructureViewTreeElement,
+    ItemPresentation,
+    NavigationItem {
+  override fun getValue() = element
+  override fun getChildren() = when (element) {
+    is SqlDelightFile -> element.childIdentifiers()
+    else -> emptyArray()
+  }
+
+  override fun canNavigateToSource() = (element as Navigatable).canNavigateToSource()
+  override fun canNavigate() = (element as Navigatable).canNavigate()
+  override fun navigate(requestFocus: Boolean) = (element as Navigatable).navigate(requestFocus)
+  override fun getPresentation() = this
+  override fun getName(): String? = when {
+    element is SqlDelightFile -> element.virtualFile?.name
+    element.node.elementType == SqliteTypes.TABLE_NAME -> "CREATE TABLE ${element.text}"
+    element.node.elementType == SqliteTypes.IDENTIFIER -> element.text
+    else -> throw IllegalStateException("Unhandled element $element")
+  }
+
+  override fun getLocationString(): String? = null
+  override fun getIcon(open: Boolean): Icon? = element.getIcon(Iconable.ICON_FLAG_READ_STATUS)
+
+  override fun getPresentableText() = name
+
+  companion object {
+    private fun SqlDelightFile.childIdentifiers(): Array<TreeElement> {
+      val tableNames = childrenOfType(SqlDelightTypes.SQL_STMT_LIST)
```

Comment:
this can be type safe using the bnf-generated psi.

```
val statementList = PsiTreeUtil.getChildOfType(this, SqlDelightSqlStmtList::class.java)!!
val tableNames = statementList.statementList.flatMap { it.sqlStmt.createTable.tableName }
val stmtIdentifiers = statement.stmtIdentifierList.map { it.childOfType(SqliteTypes.IDENTIFIER) }
```

then you dont need extension function